File: src/Ncp.Mom.Domain/AggregatesModel/WorkOrderAggregate/WorkOrder.cs
using Ncp.Mom.Domain.DomainEvents;
using Ncp.Mom.Domain.AggregatesModel.ProductAggregate;
using Ncp.Mom.Domain.AggregatesModel.ProductionPlanAggregate;
using Ncp.Mom.Domain.AggregatesModel.RoutingAggregate;

namespace Ncp.Mom.Domain.AggregatesModel.WorkOrderAggregate;

public partial record WorkOrderId : IGuidStronglyTypedId;

/// <summary>
/// 工单聚合根
/// </summary>
public partial class WorkOrder : Entity<WorkOrderId>, IAggregateRoot
{
    protected WorkOrder() { }

    public WorkOrder(
        string workOrderNumber,
        ProductionPlanId productionPlanId,
        ProductId productId,
        int quantity,
        RoutingId routingId)
    {
        WorkOrderNumber = workOrderNumber;
        ProductionPlanId = productionPlanId;
        ProductId = productId;
        Quantity = quantity;
        RoutingId = routingId;
        Status = WorkOrderStatus.Created;
        CompletedQuantity = 0;
        AddDomainEvent(new WorkOrderCreatedDomainEvent(this));
    }

    public string WorkOrderNumber { get; private set; } = string.Empty;
    public ProductionPlanId ProductionPlanId { get; private set; }= default!;
    public ProductId ProductId { get; private set; } = default!;
    public int Quantity { get; private set; }
    public int CompletedQuantity { get; private set; }
    public RoutingId RoutingId { get; private set; } = default!;
    public WorkOrderStatus Status { get; private set; }
    public DateTime? StartTime { get; private set; }
    public DateTime? EndTime { get; private set; }
    public RowVersion RowVersion { get; private set; } = new RowVersion();
    public UpdateTime UpdateTime { get; private set; } = new UpdateTime(DateTimeOffset.UtcNow);

    public void Start()
    {
        if (Status != WorkOrderStatus.Created && Status != WorkOrderStatus.Paused)
            throw new KnownException("只能启动已创建或已暂停的工单");

        Status = WorkOrderStatus.InProgress;
        StartTime = DateTimeOffset.UtcNow.DateTime;
        AddDomainEvent(new WorkOrderStartedDomainEvent(this));
    }

    public void Pause()
    {
        if (Status != WorkOrderStatus.InProgress)
            throw new KnownException("只能暂停进行中的工单");

        Status = WorkOrderStatus.Paused;
        AddDomainEvent(new WorkOrderPausedDomainEvent(this));
    }

    public void Resume()
    {
        if (Status != WorkOrderStatus.Paused)
            throw new KnownException("只能恢复已暂停的工单");

        Status = WorkOrderStatus.InProgress;
        AddDomainEvent(new WorkOrderResumedDomainEvent(this));
    }

    public void ReportProgress(int quantity)
    {
        if (Status != WorkOrderStatus.InProgress)
            throw new KnownException("只能报工进行中的工单");

        if (quantity <= 0)
            throw new KnownException("报工数量必须大于0");

        if (CompletedQuantity + quantity > Quantity)
            throw new KnownException("报工数量不能超过工单数量");

        CompletedQuantity += quantity;

        if (CompletedQuantity >= Quantity)
        {
            Status = WorkOrderStatus.Completed;
            EndTime = DateTimeOffset.UtcNow.DateTime;
            AddDomainEvent(new WorkOrderCompletedDomainEvent(this));
        }
        else
        {
            AddDomainEvent(new WorkOrderProgressReportedDomainEvent(this, quantity));
        }
    }

    public void Cancel()
    {
        if (Status == WorkOrderStatus.Completed)
            throw new KnownException("已完成的工单不能取消");

        Status = WorkOrderStatus.Cancelled;
        AddDomainEvent(new WorkOrderCancelledDomainEvent(this));
    }
}

public enum WorkOrderStatus
{
    Created,    // 已创建
    InProgress, // 进行中
    Paused,     // 已暂停
    Completed,  // 已完成
    Cancelled   // 已取消
}

File: frontend/src/api/workOrder.ts
import api from './index'
import type { PaginationRequest, PaginationResponse } from '@/types'

// 工单状态
export enum WorkOrderStatus {
  Created = 0,
  InProgress = 1,
  Paused = 2,
  Completed = 3,
  Cancelled = 4
}

// 工单类型定义
export interface WorkOrder {
  id: string
  workOrderNumber: string
  productionPlanId: string
  productId: string
  quantity: number
  completedQuantity: number
  routingId: string
  status: WorkOrderStatus
  startTime?: string
  endTime?: string
  updateTime: string
}

// 工单查询请求
export interface WorkOrderQueryRequest extends PaginationRequest {
  productionPlanId?: string
  status?: WorkOrderStatus
}

// 创建工单请求
export interface CreateWorkOrderRequest {
  workOrderNumber: string
  productionPlanId: string
  productId: string
  quantity: number
  routingId: string
}

// 报工请求
export interface ReportProgressRequest {
  workOrderId: string
  quantity: number
}

// 工单API
export const workOrderApi = {
  // 获取工单列表
  getWorkOrders: (params: WorkOrderQueryRequest) =>
    api.get<PaginationResponse<WorkOrder>>('/work-orders', { params }),

  // 获取工单详情
  getWorkOrder: (id: string) =>
    api.get<WorkOrder>(`/work-orders/${id}`),

  // 创建工单
  createWorkOrder: (data: CreateWorkOrderRequest) =>
    api.post<{ id: string }>('/work-orders', data),

  // 启动工单
  startWorkOrder: (id: string) =>
    api.post(`/work-orders/${id}/start`),

  // 暂停工单
  pauseWorkOrder: (id: string) =>
    api.post<boolean>(`/work-orders/${id}/pause`),

  // 恢复工单
  resumeWorkOrder: (id: string) =>
    api.post<boolean>(`/work-orders/${id}/resume`),

  // 取消工单
  cancelWorkOrder: (id: string) =>
    api.post<boolean>(`/work-orders/${id}/cancel`),

  // 报工
  reportProgress: (data: ReportProgressRequest) =>
    api.post(`/work-orders/${data.workOrderId}/report-progress`, { quantity: data.quantity })
}

